### PR TITLE
API(4) Move HelAmps to the latest MemoryAccess classes

### PR DESCRIPTION
**Pull request by valassi**
I have decided to strip off what was initially "the second part of apirambo" PR #321.

Essentially in apirambo I have the new MemoryAccess for rambo, the old MemoryAcccess for MEs. This new PR is about porting the new MemoryAccess to MEs and specifically HELAmps, too.

This is WIP
- I have only prototyped one function imzxxx, all other functions must be done
- The SIMD/512y performance in eemumu of the new memory access gives a 10% performance hit, to be understood
- The functional and performance tests however succeed, all segfaults and failures are fixed

What remans to be done
- fix the performance in eemumu
- port all other ixxx/oxxx functions
- rerun the throuput test in eemumu and check all is ok
- remove neppM from mgOnGpuConfig, now it will be enough to encapsulate it in MemoryAccessMomenta
- strip off p4type from MemryAccess.h, rename it as MemoryAccessVectors.h (just the fptype/fptye_v conversions)
- cleanup, remove all unused code (eg Memory.h, possibly more)
- backport to code generate
- rerun tests for ggtt and ggttgg, check performance

### CODE REVIEW

**Comment by valassi** [comment]
This is now complete.

This is where I stripped off what was initially "the second part of apirambo" PR #321. Essentially in apirambo I have the new MemoryAccessMomenta for rambo, the old MemoryAcccessMomenta for MEs in HelAmps. This new PR is about porting the new MemoryAccessMomenta to HelAmps, too. Note that new BufferMEs and MemoryAccessMEs will be created in the upcoming apimes PR.

With respect to the previous WIP
-    I had only prototyped one function imzxxx, now also all other functions have been done. Note that the choice of imz was bad: there was a functional bug and I was reading always the same first event, but as the initial momenta are the same for all events, this bug had gone unnoticed (the avereage ME was ok).
-    The SIMD/512y performance in eemumu of the new memory access was giving a 10% performance hit in eemumu. After fixing the functional bug, this disappeared and 'magically' I even got 5-10% better performance in eemumu. I have just run the ggtt and ggttgg tests however and the ggtt performance seems worse (to be understood?), especially with hardcoded parameters(?!), and only in avx2 or avx512, while none/sse4 are faster(?!). For ggttgg however all is ok.

With respect to the other points that I mentioned to be done
-    fix the performance in eemumu: done, with caveats as above for ggtt
-    port all other ixxx/oxxx functions: done
-    rerun the throuput test in eemumu and check all is ok: done
-    remove neppM from mgOnGpuConfig, now it will be enough to encapsulate it in MemoryAccessMomenta: done
-    strip off p4type from MemryAccess.h, rename it as MemoryAccessVectors.h (just the fptype/fptye_v conversions): done
-    cleanup, remove all unused code (eg Memory.h, possibly more): done, but note that Memory.h will need the apimes PR to be removed
-    backport to code generate: done
-    rerun tests for ggtt and ggttgg, check performance: done

In summary, pending:
- apimes PR (upcoming "API5"): new Buffer and MemoryAccess for MEs, also get rid of Memory.h
- understand better ggtt performance from these changes

**Comment by valassi** [comment]
This completes the "API4" step described in #323

**Comment by valassi** [comment]
This was presented at the meeting today
https://indico.cern.ch/event/1103955/
I am merging this now
